### PR TITLE
12670 - Prevent essential components being accidentally deleted

### PR DESCRIPTION
**Pull request by BrentEaston**

### CODE REVIEW

**Comment by uckelman** [comment]
I don't see anything here that would be backwards incompatible. Could this be in 3.7?

**Comment by BrentEaston** [comment]
I'm not actually very happy with this, I am going to redo it, back to the Buildable level and implement a set of new default methods isMandatory() = false, isMovable() = true and isUnique() = false so all Buildables can register themselves as  to how they want to be treated by the ConfigureTree.

And yes. it will be 3.7 compatible.

**Comment by uckelman** [comment]
Is this the way you want it to be now?

**Comment by BrentEaston** [comment]
Yes, this is good to go. More widespread changes, but it is actually much simpler and supports custom components.

File: vassal-app/src/main/java/VASSAL/configure/ConfigureTree.java [hunk 1]
```diff
@@ -1551,7 +1563,33 @@ public DefaultMutableTreeNode getTreeNode(Configurable target) {
...
+  public boolean isMoveAllowed(Configurable target) {
+    return target.isMovable();
+  }
+
+  /**
+   * Is the Component allowed to be deleted?
+   * A mandatory component can be deleted if there is more than one of them with the same parent
+   * */
+  protected boolean isDeleteAllowed(Configurable target) {
+    if (target instanceof AbstractBuildable) {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/configure/ConfigureTree.java:1576]: Could you simplify this by checking `target.isMandatory()` right at the start and returning `true` if not?

Comment (uckelman) [reply vassal-app/src/main/java/VASSAL/configure/ConfigureTree.java:1576]: I think this?
```
if (!target.isMandatory()) {
  return true;
}

if (target instanceof AbstractBuildable) {
  final Buildable parent = ((AbstractBuildable) target).getAncestor();
  if (parent instanceof AbstractBuildable) {
    final int count = ((AbstractBuildable) parent).getComponentsOf(target.getClass()).size();
    return count > 1;
  }
}

return false; 
```
Maybe that's not really simpler?

Comment (BrentEaston) [reply vassal-app/src/main/java/VASSAL/configure/ConfigureTree.java:1576]: Slightly more efficient though :) Change made.

**Pull request merged**